Add all the files which pytype fails on during import.

x3k6a2 opened the pull request:
All of these files can not be used with pytype at HEAD. They fail during import with various error messages.

matthiaskramm commented on tests/pytype_blacklist.txt:
```diff
@@ -1,13 +1,117 @@
+stdlib/3/io.pyi
+stdlib/3/calendar.pyi
+stdlib/3/fcntl.pyi
+stdlib/3/gzip.pyi
```
How come we didn't have to blacklist these before? The existing travis tests *do* check the pyi parsing step.

JelleZijlstra replied:
That particular error is because of the `' '` default, right? That seems easy to fix at least.

x3k6a2 replied:
Created https://github.com/python/typeshed/pull/1721 to cahnge the defaults.

matthiaskramm replied:
We don't actually want to use typeshed's `builtins.pyi` with pytype, though, since we have our own (`__bulitin__.pytd`). We're probably just missing a symlink `builtins.pytd -> __builtin__.pytd` in `pytype/pytd/builtins`.

x3k6a2 replied:
It becomes a BadDependency Error once builtins.pyi is changed. As the pytype typing module does not have all the things needed.

gvanrossum commented:
You can fix that in builtins.pyi though. Replace ' ' with ...

On Nov 8, 2017 7:58 AM, "Sebastian Steenbuck" <notifications@github.com>
wrote:

> *@x3k6a2* commented on this pull request.
> ------------------------------
>
> In tests/pytype_blacklist.txt
> <https://github.com/python/typeshed/pull/1720#discussion_r149711723>:
>
> >  # pytype doesn't yet support aliases with implicit type parameters
>  # (e.g., here, FutureT = Future[T])
>  stdlib/3.4/asyncio/tasks.pyi
> +
> +# ParseError: syntax error, unexpected ')', expecting NAME or NUMBER or
> +# ELLIPSIS
> +stdlib/2and3/locale.pyi
> +stdlib/3/io.pyi
> +stdlib/3/calendar.pyi
> +stdlib/3/fcntl.pyi
> +stdlib/3/gzip.pyi
>
> This change blacklists all files which will kill PyType if actually used.
> The above listed files depend on stdlib/3/builtins.pyi which contains line
> 241:
> "def center(self, width: int, fillchar: str = ' ') -> str: ..."
>
> ParseError: File: "pytype/typeshed/stdlib/3/builtins.pyi", line 241
> def center(self, width: int, fillchar: str = ' ') -> str: ...
> ^
> ParseError: syntax error, unexpected ')', expecting NAME or NUMBER or
> ELLIPSIS
>
> The travis tests just try to parse each file in isolation and can
> therefore not see this issue.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/python/typeshed/pull/1720#discussion_r149711723>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACwrMlaU0IhWl3C7XY-ogUbaI4KCJSxiks5s0c-jgaJpZM4QWVW->
> .
>

gvanrossum commented:
Oh, BTW, we should probably get rid of symlinks, not add more. They don't
work on Windows (at least, not without Administrator privileges).
See https://github.com/python/mypy/issues/4143#issuecomment-339795938

JelleZijlstra commented on tests/pytype_blacklist.txt:
```diff
@@ -1,13 +1,116 @@
-# Pytype blacklist. Files will not be tested with pytype.
+# Pytype blacklist. Files will not be tested with pytype. Notice that the parse
+# test will work for some of the files. Loading the files will fail, but is not
```
Is it possible to make typeshed's pytype tests also load these files? It's not ideal if the typeshed project has to maintain a blacklist file without a test to ensure the blacklist is correct.

x3k6a2 replied:
I agree. This will take a few more days. The code path to easily test this is not yet in PyType's github repo.

x3k6a2 commented:
I updated the pytype_test script to use pytype instead of pytd and updated the blacklist accordingly.

This is not a perfect change, it would be nicer to actually test with Python 3.6 for the stdlib/3.X files but that is too much for this change and should be done in a different one.

x3k6a2 commented:
A green travis run is at https://travis-ci.org/x3k6a2/typeshed/builds/300136808 . Afaik I can't request a rerun for the failed integration check.
This works now because of the pytype commit https://github.com/google/pytype/commit/0964db4d6683324035486109a5ea6c210d8b131c

JelleZijlstra commented:
Just restarted the Travis build.

x3k6a2 commented:
As discussed offline:
1: Added a "# parse only" comment to the 3.x files in the blacklist. Which will make those files be parsed but not loaded.
2: Checked all the stdlib/2.x/ files with exceptions. Most of the exceptions could be removed by providing the --module-name=foo flag. Without it wildcard imports are not handled correctly. The alias for ".*" is not removed by the LookupExternalTypes.VisitTypeDevlUnit visitor. This behavior is a bit surprising.
The only 2.x files added left to be added to the blacklist is os.__init__.pyi which imports builtins. Builtins is not defined for python2.

matthiaskramm commented on tests/pytype_test.py:
```diff
@@ -25,35 +34,48 @@ def main():
+        print('--- nothing to do; exit 1 ---')
         sys.exit(1)
 
+_PARSE_ONLY_REGEX = r'\s*#\s*parse only\s*'
```
Do the re.compile right here?

x3k6a2 replied:
Obsolete, done in another place.

matthiaskramm commented on tests/pytype_test.py:
```diff
@@ -25,35 +34,48 @@ def main():
+    parse_only_re = re.compile(_PARSE_ONLY_REGEX)
+    for f in lines:
+        if parse_only_re.search(f):
+            parse_only.append(f.split('#')[0].strip())
```
Make the regexp so that you can do

```
match = parse_only_re.match(f)
if match:
  parse_only.append(match.group(1))
```
?

x3k6a2 replied:
Refactored so that the regexes are independent of each other.

matthiaskramm commented on tests/pytype_test.py:
```diff
@@ -65,27 +87,49 @@ def communicate(self):
-    wanted = re.compile(r"stdlib/.*\.pyi$")
-    skipped = re.compile("(%s)$" % "|".join(load_blacklist()))
-    files = []
+    blacklist, parse_only = load_blacklist()
```
"parse_only" is from the blacklist, too, so that variable name is a bit confusing.

Rename `blacklist` to `skip`?

x3k6a2 replied:
Done

matthiaskramm commented on tests/pytype_test.py:
```diff
@@ -65,27 +87,49 @@ def communicate(self):
+                    ['pytype',
+                     '--typeshed-location=%s' % os.getcwd(),
+                     '--module-name=%s' % _get_module_name(f),
+                     '--convert-to-pickle=%s' % temp.name,
```
Do you need the temporary file? Or could you just use `os.devnull`?

x3k6a2 replied:
Done

matthiaskramm commented on tests/pytype_test.py:
```diff
@@ -97,13 +141,16 @@ def pytype_test(args):
+    print('Ran pytype with %d pyis, got %d errors.' % (runs, errors))
     return max_code, runs
 
+def _get_module_name(filename):
```
Could you move this above the location where it's called?

x3k6a2 replied:
Done
